### PR TITLE
ENH: Add functionality to hash Python dependencies through their version

### PR DESCRIPTION
**Pull request by mbaak**
ENH: Add functionality to hash Python dependencies through their version
    
- new class to PythonDependencyHash
- added related unit tests
- update of readme with dependency hash example
- Added documentation on Python package dependencies

### CODE REVIEW

File: tests/dependency/test_dependency.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+
+def test_dependency():
+    expect = {"pycodehash": pycodehash.__version__}
+    inlined_expect = inline_metadata(expect)
+    expect_hash = hash_string(inlined_expect)
+
+    hasher = DependencyHash()
+
+    result = DependencyHash.collect_metadata(["pycodehash"])
```
Comment (sbrugman) [inline tests/dependency/test_dependency.py:14]: `hasher.collect_metadata`

File: src/pycodehash/dependency/dependency.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import sys
```
Comment (sbrugman) [inline src/pycodehash/dependency/dependency.py:1]: `from __future__ import annotations`

File: README.md [hunk 3]
```diff
@@ -113,6 +113,32 @@ print(len(dh.collect_partitions(Path(__file__).parent / "src")))
 ```
 _[Dataset Usage Example](https://github.com/pycodehash/pycodehash/blob/main/example_data.py)_
 
+
+### Dependencies
+
```
Comment (sbrugman) [inline README.md:118]: Would be good to also add a page in the docs and a link in the section `Detecting changes in data pipelines`

File: src/pycodehash/dependency/dependency.py [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+
+class DependencyHash(ApproximateHasher):
+    """Get the hash of a list of user-provided dependencies """
+    @staticmethod
+    def collect_metadata(dependencies: list[str]) -> dict[str, str]:
+        for dep in dependencies:
+            if dep not in sys.modules:
+                raise ModuleNotFoundError(f'No module named "{dep}"')
+        return {dep: version(dep) for dep in sorted(dependencies)}
```
Comment (RUrlus) [inline src/pycodehash/dependency/dependency.py:14]: What does version do when there is no version that can be found? Packages always have versions?

File: src/pycodehash/dependency/dependency.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
+import sys
+from importlib.metadata import version
+
+from pycodehash.datasets.approximate_hasher import ApproximateHasher
+
+
+class DependencyHash(ApproximateHasher):
+    """Get the hash of a list of user-provided dependencies """
+    @staticmethod
+    def collect_metadata(dependencies: list[str]) -> dict[str, str]:
```
Comment (RUrlus) [inline src/pycodehash/dependency/dependency.py:10]: Should we include the std lib by default?

File: README.md [hunk 6]
```diff
@@ -20,6 +20,7 @@ Learn more on how PyCodeHash detects changes in:
 - [Python Functions](https://pycodehash.github.io/pycodehash/python_functions/)
 - [SQL Queries](https://pycodehash.github.io/pycodehash/sql_queries/)
 - [Datasets](https://pycodehash.github.io/pycodehash/datasets/): Files, Directories, S3, Hive
+- [Dependencies](https://pycodehash.github.io/pycodehash/dependencies/)
```
Comment (sbrugman) [inline README.md:23]: Dependencies.md is missing

File: src/pycodehash/dependency/dependency.py (hunk 4 above)
Comment (mbaak) [reply src/pycodehash/dependency/dependency.py]: I think it's safe. From what I see the version is extracted from the package wheel. And the PEP 491 standard (The Wheel Binary Package Format) requires one. (The version source code assumes the version is present in the metadata.)

**Comment by RUrlus** [comment]
Final note @mbaak. Can you rename the PR according to something like:
`ENH: Add functionality to hash Python dependencies through their version`

File: README.md (hunk 6 above)
Comment (mbaak) [reply README.md]: Added

File: tests/dependency/test_dependency.py [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+    inlined_expect = inline_metadata(expect)
+    expect_hash = hash_string(inlined_expect)
+
+    hasher = PythonDependencyHash()
+
+    result = hasher.collect_metadata(["pycodehash"])
+    assert result == expect
+
+    result_hash = hasher.compute_hash(["pycodehash"])
+    assert result_hash == expect_hash
```
Comment (sbrugman) [inline tests/dependency/test_dependency.py:18]: We can test for add Python version (store the hashes from 3.8 to 3.11)

Comment (mbaak) [reply tests/dependency/test_dependency.py:18]: Ok, will add.

Comment (mbaak) [reply tests/dependency/test_dependency.py:20]: Added.

**Review by sbrugman — APPROVED** [review]

**Pull request merged**